### PR TITLE
App Start Ending Early Bug Fix

### PR DESCRIPTION
**Pull request by ssakhamu**
Fixes app start root span ending early bug by: 

- creating a check for a boolean isDocumentLoaded on root span end
- quiescing app start root span if isDocumentLoaded is not true
- setting context for UI render start to be the active context to avoid using stale root spans as context
- adding to existing test a check prior to the start of doc-load span to test early end edge case

### CODE REVIEW

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 1]
```diff
@@ -130,13 +129,12 @@ class UiRenderSpanManager {
...
 
   private startUiRenderSpan(startTimeMs: number = Date.now()): void {
-    const parentContext = trace.setSpan(context.active(), this.parentSpan);
     this.activeRenderSpan = this.tracer.startSpan(
       'UI Render',
       {
         startTime: startTimeMs
       },
-      parentContext
+      context.active()
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts:137]: nice simplification!

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 2]
```diff
@@ -155,33 +153,42 @@ class UiRenderSpanManager {
 class ResourceFetchSpans {
   private activeSpanIds: Set<string>;
   private lastSpanCompletedAtMs?: number;
+  private isDocumentLoaded: boolean;
 
-  constructor() {
+  constructor(isAppStart: boolean = false) {
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: Why not just have the param directly mirror the value it's being set to (e.g. `constructor(isDocumentLoaded: boolean = true)`?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 3]
```diff
@@ -155,33 +153,42 @@ class UiRenderSpanManager {
...
   }
 
+  setLastSpanCompletedAtMs(completedAtMs: number): void {
+    if (
+      this.lastSpanCompletedAtMs === undefined ||
+      completedAtMs > this.lastSpanCompletedAtMs
+    ) {
+      this.lastSpanCompletedAtMs = completedAtMs;
+    }
+  }
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts:177]: Ooh was this also a bug? I'd imagine that without this logic, a later span reporting it's earlier end time could cause the root span to end artificially early.

Can you add a test case for this?

**Review by rebehe — APPROVED** [review]

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 2 above)
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: don't need the type here since isAppState = false implies the type (go/tsjs-style#type-inference)

Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: why have two different booleans here? it seems like there are two paths centered around if "isAppStartRootSpan" or if it's not

**Review by MaesterChestnut — APPROVED** [review]

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 2 above)
Comment (rebehe) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts]: TIL

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 4]
```diff
@@ -130,13 +129,12 @@ class UiRenderSpanManager {
...
   private startUiRenderSpan(startTimeMs: number = Date.now()): void {
-    const parentContext = trace.setSpan(context.active(), this.parentSpan);
     this.activeRenderSpan = this.tracer.startSpan(
       'UI Render',
       {
         startTime: startTimeMs
       },
-      parentContext
+      context.active()
     );
```
[the remark this answers is not in the document]
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:138]: - the root span may be stale at the time of setting parentContext
- I don't think there is a clear specification in the design for a UI render span to only exist as a direct child to the root span (seems against OTEL convention to change context)

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 5]
```diff
@@ -155,33 +153,42 @@ class UiRenderSpanManager {
...
   private activeSpanIds: Set<string>;
   private lastSpanCompletedAtMs?: number;
+  private isDocumentLoaded: boolean;
 
-  constructor() {
+  constructor(isAppStart: boolean = false) {
     this.activeSpanIds = new Set<string>();
     this.lastSpanCompletedAtMs = undefined;
+    this.isDocumentLoaded = !isAppStart;
   }
```
[the remark this answers is not in the document]
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts]: Seems like new logic that is different from the original code. Can adopt this only if this edge case ever arises: [go/tott/737](http://goto.google.com/tott/737).
